federation-js/src/service/printSupergraphSdl.ts

```diff
@@ -375,54 +384,83 @@ function printJoinFieldDirectives(
-    // look at the owning service here, that meant we would print `@join__field`
-    // without a corresponding `@join__type`, which is invalid according to the spec.
-    const { serviceName, keys } = parentType.extensions?.federation;
-    if (serviceName && keys) {
-      const enumValue = context.graphNameToEnumValueName?.[serviceName];
-      assert(
-        enumValue,
-        `Unexpected enum value missing for subgraph ${serviceName}`,
-      );
-      return printed + `graph: ${enumValue})`;
-    }
-    return '';
+  const directiveArgs: string[] = [];
+
+  const fieldMetadata: FederationField | undefined =
+    field.extensions?.federation;
+
+  let serviceName = fieldMetadata?.serviceName;
+
+  // For entities (which we detect through the existence of `keys`),
+  // although the join spec doesn't currently require `@join__field(graph:)` when
+  // a field can be resolved from the owning service, the code we used
+  // previously did include it in those cases. And especially since we want to
+  // remove type ownership, I think it makes to keep the same behavior.
+  if (!serviceName && parentType.extensions?.federation.keys) {
+    serviceName = parentType.extensions?.federation.serviceName;
   }
 
-  const {
-    serviceName,
-    requires = [],
-    provides = [],
-  }: FederationField = field.extensions.federation;
+  if (serviceName) {
+    const enumValue = context.graphNameToEnumValueName?.[serviceName];
+    assert(
+      enumValue,
+      `Unexpected enum value missing for subgraph ${serviceName}`,
+    );
+
+    directiveArgs.push(`graph: ${enumValue}`);
+  }
 
-  let directiveArgs: string[] = [];
+  const requires = fieldMetadata?.requires;
 
-  if (serviceName && serviceName.length > 0) {
+  if (requires && requires.length > 0) {
     directiveArgs.push(
-      `graph: ${context.graphNameToEnumValueName?.[serviceName] ?? serviceName}`,
+      `requires: ${printStringLiteral(printFieldSet(requires))}`,
     );
   }
 
-  if (requires.length > 0) {
-    directiveArgs.push(`requires: ${printStringLiteral(printFieldSet(requires))}`);
-  }
+  const provides = fieldMetadata?.provides;
 
-  if (provides.length > 0) {
-    directiveArgs.push(`provides: ${printStringLiteral(printFieldSet(provides))}`);
+  if (provides && provides.length > 0) {
+    directiveArgs.push(
+      `provides: ${printStringLiteral(printFieldSet(provides))}`,
+    );
   }
 
-  printed += directiveArgs.join(', ');
+  // A directive without arguments isn't valid (nor useful).
+  if (directiveArgs.length < 1) return '';
+
+  return ` @join__field(${directiveArgs.join(', ')})`;
+}
 
-  return (printed += ')');
+function printAppliedDirectives(field: GraphQLField<any, any>) {
+  const appliedDirectives = (
+    field.extensions?.federation?.appliedDirectives ?? []
+  ) as DirectiveNode[];
+  return appliedDirectives.length > 0
+    ? ' ' + appliedDirectives.map(printDirectiveNode).join(' ')
+    : '';
+};
+
+function printDirectiveNode(directive: DirectiveNode) {
```

Comment:
I think this could just be `print` from `graphql-js`? That would also make sure argument types are handled consistently.

Reply:
Why didn't I think of that? 😄
86e4ed2b